### PR TITLE
[CMD][BOOTDATA] Minor improvements

### PR DESCRIPTION
**Pull request by cbialorucki**
## Purpose

Jira issues [CORE-16193](https://jira.reactos.org/browse/CORE-16193), [CORE-17031](https://jira.reactos.org/browse/CORE-17031)

To improve the command prompt experience in ReactOS.

## Proposed changes

- Remove a hardcoded copyright string and move into localizable resources.
- Remove the `PROMPT` environment variable from clean installs of ReactOS. By default, the command prompt uses `$P$G` (path + '>') as its prompt settings and does not require this environment variable. Clean installs of Windows Server 2003 do not include this environment variable either. I documented this environment variable in our [wiki](https://reactos.org/wiki/Prompt_(environment_variable)) if anyone would like to set it on their own ReactOS installs.
- Remove the new line above the copyright notice when the information line is turned off.

## TODO (in future PRs)

- Add a new environment variable to customize the information line.

### CODE REVIEW

**Comment by HBelusca** [comment]
The /I prompt line was added by me as hommage to OS/2 and as a base functionality to have a future user-customizable control line or something akin to that.
So, I would instead recommend you just remove the prompt environment variable from the registry, and keep your localization of the copyright string, without touching the rest. Thanks.

File: base/shell/cmd/cmd.c [hunk 1]
```diff
@@ -2299,11 +2299,10 @@ Initialize(VOID)
     if (!*ptr)
     {
         /* If neither /C or /K was given, display a simple version string */
-        ConOutChar(_T('\n'));
         ConOutResPrintf(STRING_REACTOS_VERSION,
                         _T(KERNEL_VERSION_STR),
                         _T(KERNEL_VERSION_BUILD_STR));
-        ConOutPuts(_T("(C) Copyright 1998-") _T(COPYRIGHT_YEAR) _T(" ReactOS Team.\n"));
+        ConOutResPrintf(STRING_CMD_COPYRIGHT, _T(COPYRIGHT_YEAR));
```
Comment (HBelusca) [inline base/shell/cmd/cmd.c:2305]: Notice also there: https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/cmd/ver.c;h=7d4e497688db763eca57fcf0823bd6edfb794ba2;hb=HEAD#l143

**Comment by cbialorucki** [comment]
> The /I prompt line was added by me as hommage to OS/2 and as a base functionality to have a future user-customizable control line or something akin to that.

We can move the information line setting to another part of the registry and handle it entirely in `cmd.c`. Then draw it just before the copyright notice. Should be more straightforward that way.

File: base/shell/cmd/lang/cs-CZ.rc [hunk 2]
```diff
@@ -531,7 +531,7 @@ title         new title\n"
     STRING_REPLACE_ERROR5 "Přístup odepřen - %s\n"
     STRING_REPLACE_ERROR6 "Nebyly nazeleny žádné soubory - %s\n"
     STRING_REPLACE_ERROR7 "Rozšířená chyba 32\n"
-    STRING_CMD_INFOLINE "  ReactOS Příkazový řádek                              Napsání HELP = Nápověda  "
+    STRING_CMD_COPYRIGHT "(C) Copyright 1998-%s ReactOS Team.\n\n"
```
Comment (JoachimHenze) [inline base/shell/cmd/lang/cs-CZ.rc:534]: I do like your PR, but I would suggest to use the year 1996 here instead of 1998 (same as used in our shell32, and same as in our pictures for the session startup/shutdown.

**Review by JoachimHenze — APPROVED** [review]

**Comment by JoachimHenze** [comment]
For the record: I am pretty sure that your PR will fix https://jira.reactos.org/browse/CORE-16193 and most likely also the (duplicate) https://jira.reactos.org/browse/CORE-17031
so it makes sense to mention that in the PRs description!

File: base/shell/cmd/cmd.c [hunk 3]
```diff
@@ -2299,11 +2346,20 @@ Initialize(VOID)
     if (!*ptr)
     {
         /* If neither /C or /K was given, display a simple version string */
-        ConOutChar(_T('\n'));
+
+        /* Print the information bar if its environment variable is set. */
+        if (GetEnvironmentVariable(_T("INFOLINE"), szInfoBar, _countof(szInfoBar))  &&
```
Comment (binarymaster) [inline base/shell/cmd/cmd.c:2351]:
```suggestion
        if (GetEnvironmentVariable(_T("INFOLINE"), szInfoBar, _countof(szInfoBar)) &&
```

File: base/shell/cmd/cmd.c [hunk 4]
```diff
@@ -2124,6 +2124,52 @@ GetCmdLineCommand(
...
+    /* Return directly if the output handle is not a console handle */
+    if (!GetConsoleScreenBufferInfo(hOutput, &csbi))
+        return;
+
+    if (wcscmp(InfoBar, _T("1")) == 0)
+    {
+        iInfoLineLen = LoadString(CMD_ModuleHandle, STRING_CMD_INFOLINE, (PTSTR)&pszInfoLine, 0);
+    }
+
+    else
```
Comment (binarymaster) [inline base/shell/cmd/cmd.c:2151]:
```suggestion
    else
```

File: base/shell/cmd/cmd.c [hunk 5]
```diff
@@ -2299,11 +2345,20 @@ Initialize(VOID)
     if (!*ptr)
     {
         /* If neither /C or /K was given, display a simple version string */
-        ConOutChar(_T('\n'));
+
+        /* Print the information bar if its environment variable is set. */
```
Comment (HBelusca) [inline base/shell/cmd/cmd.c:2349]: Please revert these changes. If the info line is printed when the prompt is shown, this is because when the console scrolls past one screenful, the info line would be lost if it is printed only once as you do it there now.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: base/shell/cmd/cmd.c (hunk 5 above)
Comment (cbialorucki) [reply base/shell/cmd/cmd.c:2349]: The new line above the copyright notice is a hack.

Comment (HBelusca) [reply base/shell/cmd/cmd.c:2349]: > the top of the window after the prompt, but rather at the top of the entire console history

It is redrawn at the top of the **console screenbuffer** as it should be. (I tested on OS/2 etc.)

Comment (HBelusca) [reply base/shell/cmd/cmd.c:2349]: Note that when I said "these changes" I was talking about moving the printInfoLine function to cmd.c and removing the /I prompt switch.

Comment (cbialorucki) [reply base/shell/cmd/cmd.c:2349]: Well if that's how it's supposed to work, guess that's fine. Just seemed odd to me. 

I can move the line draw code back into `prompt.c` and write a function in `cmd.c` to determine if the prompt environment variable has $I in it. Then insert a new line above the copyright notice only if it returns true.

In the future, the information line would be a lot more useful if it had its own environment variable to customize it.

Comment (HBelusca) [reply base/shell/cmd/cmd.c:2349]: > I can move the line draw code back into prompt.c and write a function to determine if the prompt environment variable has $I in it. Then insert a new line above the copyright notice only if it returns true.

Yes. That code: https://github.com/reactos/reactos/pull/5745/files#diff-499121b37bcb4ad5b9aad691293f228a3afe7d2de8e81960459033a04c9d8a21L120
(that you removed in the current state of your PR as I'm currently writing this comment) could be made into a helper function, like `BOOL IsInfoLineShown(VOID)` (or similar name), that you could call in where the extra `\n` is issued in the Initialize() function.
(And that helper also called where the original removed code was, if it still makes sense.)

> In the future, the information line would be a lot more useful if it had its own environment variable to customize it.

Still possible. Where there is this call in the PrintInfoLine function:
```c
    iInfoLineLen = LoadString(CMD_ModuleHandle, STRING_CMD_INFOLINE, (PTSTR)&pszInfoLine, 0);
```
you could do, schematically,
```c
...
    TCHAR szInfoBar[256];
...
    iInfoLineLen = GetEnvironmentVariable(_T("INFOLINE"), szInfoBar, _countof(szInfoBar));
    if (iInfoLineLen != 0 && *szInfoBar) // You should check whether *szInfoBar check is necessary in that case...
    {
        pszInfoLine = szInfoBar;
    }
    else
    {
        iInfoLineLen = LoadString(CMD_ModuleHandle, STRING_CMD_INFOLINE, (PTSTR)&pszInfoLine, 0);
    }
...
```
Also I'm not sure about the magic "0" and "1" strings you are currently using in your PR.

Comment (cbialorucki) [reply base/shell/cmd/cmd.c]: I moved the draw info line code back to `prompt.c` and added a new function to check if the info line is displayed or not.

**Comment by whindsaks** [comment]
If the PROMPT variable is going away by default, could the $I only be added automatically when there is no /C nor /K so it stops hiding the first line output?

**Comment by cbialorucki** [comment]
> If the PROMPT variable is going away by default, could the $I only be added automatically when there is no /C nor /K so it stops hiding the first line output?

The information line will not be on by default. If there is no system prompt variable, it will use $P$G.

**Comment by HBelusca** [comment]
Gotta recheck this PR tomorrow Monday :+1:

**Review by whindsaks — APPROVED** [review]

**Comment by HBelusca** [comment]
Approved post-merged :+1: :joy:
